### ISSUE TITLE
Current experiment object not working with tensorflow RunConfig is missing "environment"

### ISSUE THREAD

**Issue by zyaar** [rstudio/tfestimators#77]
Created a new experiment() and tried "train_and_evaluate()".

Got error that RunConfig is missing an 'environment' attribute.

I believe this is because you are using the newer RunConfig (tf.estimator.RunConfig) which does not include an 'environment' attribute instead of the older tf.contrib.learn.RunConfig which does, and the Experiment object still requires the older one.

**Comment by terrytangyuan** [comment]
Yes exactly. I had a PR to upstream Python interface and thought it was fixed. But let me look into this again. I recommend hold off on using experiment until the Experiment API lands in core though.

**Comment by zyaar** [comment]
Thx!

Right now am just using your package to generate a Model_fn object and then
doing everything else directly with the tensorflow package and
contrib.learn's Estimator object and it is working fine. (for some reason
the Estimator object doesn't like a Model_fn coming directly from R).

But I really the work that you are doing and would like to use it once
everything stabilizes.

- Ziv

**Comment by terrytangyuan** [comment]
Yeah I'll see what I can do there. I closed this PR I created since I thought it was fixed internally but now it seems not: https://github.com/tensorflow/tensorflow/pull/11385

**Comment by terrytangyuan** [comment]
@zyaar model_fn is a bit trickier! That's exactly where we hope this package brings to help to avoid this type of issues!

**Comment by terrytangyuan** [comment]
@zyaar A quick solution is to use `tf$contrib$learn$RunConfig()` when you initialize the estimator and then pass to experiment like follows:

```
clf <-
    linear_regressor(
      feature_columns = linear_feature_columns,
      config = tf$contrib$learn$RunConfig()
    ) %>% train(input_fn = input_fn, steps = 2)

  experiment <- experiment(
    clf,
    train_input_fn = input_fn,
    eval_input_fn = input_fn,
    train_steps = 3L,
    eval_steps = 3L,
    continuous_eval_throttle_secs = 60L
  )
train(experiment)
```
I couldn't find a good way to resolve this since [only certain existing properties of the core RunConfig are mutable](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/estimator/run_config.py#L31), e.g. you cannot add an additional field called `environment`. 

Core RunConfig and contrib RunConfig also have some fundamental differences for certain fields. 

@jjallaire @kevinushey In the meantime, do you think we should remove Experiment from public API until it's available in core, similar to other incoming canned estimators? 

Or, I can re-open this PR tensorflow/tensorflow#11385 and then give instructions to users for both v1.3.0 and dev version (with the fix) of TensorFlow?

**Comment by kevinushey** [comment]
I agree we should remove Experiment from the public API for now -- we should target whatever is considered 'stable' in v1.3 right now, and leave Experiments for v1.4 (if they make it in then).

**Comment by zyaar** [comment]
Thanks for the update.  Yes, I looked at something like this earlier.  At
the end I ended up going with a major cludge, but was able to use the
learn_runner.run() method in TF (I eventually want to play around with
learn_runner.tune() to get the hyper-parameter tuning, instead of having to
write my own).

Basically, I'm using your package to build the model_fn function (which
doesn't seem to work with pure R), and then do the rest in Tensorflow

So the following code works for me quite nicely, at least until the
Experimenter interface gets stabilized:


experiment_fn <- function(output_dir)
{
  # use tfestimators just because it can create a tensorflow compatible
model_fn function
  *model        <- estimator(fin_rnn_model$rnn_model2, model_dir =
get_most_recent_run(temp_model_store))*

  # Take the model_fn created above and create a real Estimator and
Experiment
  # object and pass back
  return(
    tf$contrib$learn$Experiment(
      tf$contrib$learn$Estimator(*model$model_fn*, model_dir =
get_most_recent_run(temp_model_store)),
      train_input_fn    = fin_rnn_model$input_train_fn,
      eval_input_fn     = fin_rnn_model$input_test_fn,
      eval_metrics      = list("rmse" =
tf$contrib$learn$MetricSpec(metric_fn =
tf$contrib$metrics$streaming_root_mean_squared_error)),
      export_strategies = NULL
    )
  )
}

...

#
# MAIN PROGRAM LOOP -----------
# /////////////////////////////

# Use learn_runner to run the experiment
tf$contrib$learn$python$learn$learn_runner$run(experiment_fn =
experiment_fn, output_dir = get_most_recent_run(temp_model_store))

On Sun, Aug 20, 2017 at 4:45 PM, Yuan (Terry) Tang <notifications@github.com

**Comment by jjallaire** [comment]
@terrytangyuan and @kevinushey We should be able to get to a place where the R model_fn is *also* the Python model_fn (i.e. it's just the reference to the reticulate Python object. I realize that there is a wrapper object there now, in Keras we don't have any such wrappers, which of course promotes interoperability with lower level Python code (as is desired here).

Let me know if there is a reason that we need the wrapper and/or if there are other reasons we can't just use the Python model_fn on the R side.

**Comment by terrytangyuan** [comment]
@jjallaire As far as I can remember, that wrapper was introduced because reticulate cannot preserve the R function signature when it's passed to another function argument. We definitely want to get rid of it though.

**Comment by terrytangyuan** [comment]
we are relying on this wrapper right now: https://github.com/rstudio/tfestimators/blob/master/inst/python/estimatortools/functions.py

**Comment by jjallaire** [comment]
That wrapper should still work fine w/ what I am talking about. 

@zyaar If you could explain a bit more narrowly what you are trying to do and what's not working hopefully we can sort it out.

**Comment by terrytangyuan** [comment]
(Please correct me if I am wrong @zyaar) I think @zyaar was trying to use raw tensorflow R package to access estimators, e.g. tf$contrib$estimator. However, the estimator Python API requires the model_fn to have the correct signature with named arguments while the model_fn defined in R being passed to estimator() lost the signature, which is where tfestimators' internal as_model_fn() brings into help - it wraps the signature correctly.

**Comment by terrytangyuan** [comment]
I think there's a miscommunication here. @zyaar didn't mean that our as_model_fn() didn't work. He just meant the opposite - it works for him and he couldn't do it without our wrapper since using raw TensorFlow with reticulate won't pass the model_fn he defined correctly. @jjallaire Does this clarify things better?

**Comment by jjallaire** [comment]
Yes, I apologize for not grasping this before! Sound like tfestimators has
got it exactly right here

**Comment by zyaar** [comment]
Yup, I think the summary in this thread is on point.  Here is the "quick"
backstory:

1.  Been working directly with the tensorflow package in R up until now
2.  Tried to create my own estimator (to use with an experiment to
ultimately use with learn_runner) with contrib.learn estimator, didn't work
because the R function that I wrote for 'model_fn' failed to get the 'mode'
argument
3.  Switched to the tf.estimator estimator and the whole thing failed
because tf.estimator does argument validation (with the contrib.learn one
didn't) and wasn't finding the arguments that it needed from the r function
(although they were already there)
4.  Some googling found your tfestimator package re-factored code to use it
instead of the raw tensorflow APIs (was very quick, great idea, I think we
need a lot more R support for tensorflow and I'm glad you guys are leading
the way)
5.  Got the estimator working fine across train, eval, infer
6.  Moved on the to experiment object, evaluator fails given that
experiment still expects an evaluator with a contrib.learn.runconfig object
which has an 'environment' attribute which the tf.estimator.runconfig
object DOES NOT.
7.  Switched back to my original tensorflow API code(still using the
contrib.learn Estimator since it works with the contrib.learn Experiment),
but used your tfestimator package to build the 'model_fn' since it seems to
be generating the correct function signature for the
contrib.learn.Estimator object to correctly populate the arguments
(specifically the mode).

Now, everything is working fine (so I'm happy), but not the ideal longer
term solution.

At the end of the day, all I really want is the ability to plug a model
into a framework and automatically build the model with alternating wave of
train and eval AND tune the hyper parameters.

I'm heading in that direction now, but if you were interested, that's my
goal.

- Ziv

**Comment by terrytangyuan** [comment]
Thanks! Back to this issue and your step 6, since it seems like
learn.RunConfig doesn't seem to work I just moved Experiment API from the
public API until it's in core. Closing this issue now. Please stay tuned!
